Documentation/coding-guidelines/interop-pinvokes.md

```diff
@@ -0,0 +1,239 @@
+
+[Default Marshalling for Boolean Types](https://msdn.microsoft.com/en-us/library/t2t3725f.aspx "MSDN")  
+
+Guids
+-----
+
+Guids are usable directly in signatures. When passed by ref, however, they should *not* be marked as `out` or `ref`. Instead the parameter should get the `[MarshalAs(UnmanagedType.LPStruct)]` attribute.
+
+| Guid | By ref Guid |
+|------|-------------|
+| `KNOWNFOLDERID` | `REFKNOWNFOLDERID` |
+
+`[MarshalAs(UnmanagedType.LPStruct)]` should _only_ be used for by ref Guids.
+
+Common Windows Data Types
+-------------------------
+
+
+|Windows        | C                 | C#    | Alternative |
+|---------------|-------------------|-------|-------------|
+|`BOOL`           |`int`                |`int`    |`bool`
+|`BOOLEAN`        |`unsigned char`      |`byte`   |`[MarshalAs(UnmanagedType.U1)] bool`
+|`BYTE`           |`unsigned char`      |`byte` | |
+|`CHAR`           |`char`               |`sbyte` | |
+|`UCHAR`          |`unsigned char`      |`byte` | |
+|`SHORT`          |`short`              |`short` | |
+|`CSHORT`         |`short`              |`short` | |
+|`USHORT`         |`unsigned short`     |`ushort` | |
+|`WORD`           |`unsigned short`     |`ushort` | |
+|`ATOM`           |`unsigned short`     |`ushort` | |
+|`INT`            |`int`                |`int` | |
+|`LONG`           |`long`               |`int` | |
+|`ULONG`          |`unsigned long`      |`uint` | |
+|`DWORD`          |`unsigned long`      |`uint` | |
+|`LARGE_INTEGER`  |`__int64`            |`long` | |
+|`LONGLONG`       |`__int64`            |`long` | |
+|`ULONGLONG`      |`unsigned __int64`   |`ulong` | |
+|`ULARGE_INTEGER` |`unsigned __int64`   |`ulong` | |
+|`UCHAR`          |`unsigned char`      |`byte` | |
+|`HRESULT`        |`long`               |`int` | |
+
+
+| Signed Pointer Types (`IntPtr`) | Unsigned Pointer Types (`UIntPtr`) |
+|----------------------------------|-------------------------------------|
+| `HANDLE` | `WPARAM` |
+| `HWND` | `UINT_PTR` |
+| `HINSTANCE` | `ULONG_PTR` |
+| `LPARAM` | `SIZE_T` |
+| `LRESULT` | |
+| `LONG_PTR` | |
+| `INT_PTR` | |
+
+[Windows Data Types](http://msdn.microsoft.com/en-us/library/aa383751.aspx "MSDN")  
+[Data Type Ranges](http://msdn.microsoft.com/en-us/library/s3f49ktz.aspx "MSDN")
+
+Blittable Types
+---------------
+Blittable types are types that have the same representation for native code. As such they do not need to be converted to another format to be marshalled to and from native code.
+
+**Blittable types:**
+
+- `byte`, `sbyte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`, `single`, `double`
+- non-nested one dimensional arrays of blittable types (e.g. `int[]`)
+- structs and classes with fixed layout that only have blittable types for instance fields
+  - fixed layout requires `[StructLayout(LayoutKind.Sequential)]` or `[StructLayout(LayoutKind.Explicit)]`
+  - structs are `LayoutKind.Sequential` by default, classes are `LayoutKind.Auto`
+
+**NOT blittable:**
+
+- `bool`
```

Comment:
Note that there is blittable version of Windows `BOOL` defined in `src\Common\src\Interop\Windows\Interop.BOOL.cs`. It is convenient to define structs that contain Windows BOOL as blittable.